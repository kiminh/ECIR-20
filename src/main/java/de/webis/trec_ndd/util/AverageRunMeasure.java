package de.webis.trec_ndd.util;

import java.util.List;
import java.util.Map;

import org.apache.commons.lang3.tuple.Pair;

import com.google.common.collect.ImmutableMap;

import lombok.Data;

@Data
public class AverageRunMeasure {
	private final List<Pair<String, Double>> originalRanking;

	public Map<String, Double> reportAvg(List<Pair<String, Double>> ranking) {
		double origAvg = avg(originalRanking);
		double newAvg = avg(ranking);
		
		return ImmutableMap.<String, Double>builder()
			.put("avgOriginal", origAvg)
			.put("avgManipulated", newAvg)
			.put("difference", origAvg - newAvg)
			.build();
	}

	private static final double avg(List<Pair<String, Double>> ranking) {
		return ranking.stream()
			.mapToDouble(Pair::getRight)
			.average().getAsDouble();
	}
}
