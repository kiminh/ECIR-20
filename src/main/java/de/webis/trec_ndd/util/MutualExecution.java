package de.webis.trec_ndd.util;

import de.webis.trec_ndd.spark.QrelConsistentMaker;
import de.webis.trec_ndd.spark.RunResultDeduplicator;

public class MutualExecution {

	public static Boolean shouldExecuteTogether(QrelConsistentMaker qrel, RunResultDeduplicator dedup) {
		return !"maxValueAllDuplicateDocsIrrelevant".equals(qrel.name()) || !"base".equals(dedup.name());
	}

}
