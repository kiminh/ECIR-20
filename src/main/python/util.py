def all_files_in_hdfs_dir(hdfs_path):
    import os
    import re
    import subprocess

    cmd = 'hdfs dfs -ls ' + hdfs_path
    cmd_output = subprocess.check_output(cmd, shell=True)
    ret = []
    for line in cmd_output.decode('utf-8').strip().split('\n'):
        fields = re.split('\s+', line)
        if len(fields) == 8:
            ret += [fields[-1]]

    return ret

