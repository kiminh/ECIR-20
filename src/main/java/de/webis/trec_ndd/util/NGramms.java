package de.webis.trec_ndd.util;

import java.io.Serializable;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.regex.Pattern;

import de.webis.trec_ndd.similarity.MD5;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

public class NGramms {

	private static final Pattern WHITESPACE = Pattern.compile("\\s+");
	
	@Data
	@NoArgsConstructor
	@AllArgsConstructor
	@SuppressWarnings("serial")
	public static class Word8Gramm implements Serializable {
		private String nGramm;
		private String md5Hash;
	}
	
	public static List<Word8Gramm> build8Gramms(String text) {
		return build8Gramms(tokenize(text));
	}
	
	public static List<String> tokenize(String text) {
		return Arrays.asList(WHITESPACE.split(text));
	}
	
	public static List<Word8Gramm> build8Gramms(List<String> tokens) {
		int nGrammLength = 8;
		List<Word8Gramm> ret = new LinkedList<>();
		
		for(int i=0; i<= tokens.size() - nGrammLength; i++) {
			ret.add(buildWord8GramStartingAt(i, tokens));
		}
		
		return ret;
	}
	
	private static Word8Gramm buildWord8GramStartingAt(int position, List<String> tokens) {
		String ret = "";
		
		for(int i = 0; i< 8; i++) {
			ret += tokens.get(position + i) + " ";
		}
		
		ret = ret.trim();
		
		return new Word8Gramm(ret, MD5.md5hash(ret));
	}
}
