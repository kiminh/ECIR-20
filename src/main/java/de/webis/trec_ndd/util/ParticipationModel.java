package de.webis.trec_ndd.util;

import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import org.apache.commons.lang3.tuple.Pair;

import static de.webis.trec_ndd.util.ScoringChanges.failIfScoresContainsDuplicatedRuns;
import static de.webis.trec_ndd.util.ScoringChanges.failIfUnknownElementsAreInRuns;
import static de.webis.trec_ndd.util.ScoringChanges.sortedStream;

public class ParticipationModel {
	private final List<Pair<String, Double>> originalScores;
	
	public ParticipationModel(List<Pair<String, Double>> originalScores) {
		failIfScoresContainsDuplicatedRuns(originalScores);
		this.originalScores = originalScores;
	}
	public List<Integer> calculateCheatingVector(List<Pair<String, Double>> manipulatedScores) {
		failIfScoresContainsDuplicatedRuns(manipulatedScores);
		failIfUnknownElementsAreInRuns(originalScores, manipulatedScores);
		
		return sortedStream(manipulatedScores)
				.map(i -> calculateCheatingScoreForEntry(i, manipulatedScores))
				.collect(Collectors.toList());
	}
	
	public List<Integer> calculateIdealVector(List<Pair<String, Double>> manipulatedScores) {
		failIfScoresContainsDuplicatedRuns(manipulatedScores);
		failIfUnknownElementsAreInRuns(originalScores, manipulatedScores);
		
		return sortedStream(originalScores)
				.map(i -> calculateIdealScoreForEntry(extractRunWithName(i.getKey(), manipulatedScores), originalScores))
				.collect(Collectors.toList());
	}

	private Integer calculateCheatingScoreForEntry(Pair<String, Double> entry, List<Pair<String, Double>> manipulatedScores) {
		String cheater = entry.getKey();
		
		Map<String, Integer> baselinePositions = runToPosition(manipulatedScores);
		List<Pair<String, Double>> runsWithCheater = new LinkedList<>(manipulatedScores).stream()
				.filter(i -> !cheater.equals(i.getKey()))
				.collect(Collectors.toList());
		runsWithCheater.add(cheaterScore(cheater));
		
		Map<String, Integer> positionsWithSingleCheater = runToPosition(runsWithCheater);
		
		return baselinePositions.get(cheater) - positionsWithSingleCheater.get(cheater);
	}
	
	private Integer calculateIdealScoreForEntry(Pair<String, Double> entry, List<Pair<String, Double>> originalScores) {
		String ideal = entry.getKey();
		Map<String, Integer> baselinePositions = runToPosition(originalScores);
		List<Pair<String, Double>> runsWithIdeal = new LinkedList<>(originalScores).stream()
				.filter(i -> !ideal.equals(i.getKey()))
				.collect(Collectors.toList());
		runsWithIdeal.add(entry);

		Map<String, Integer> positionsWithSingleIdeal = runToPosition(runsWithIdeal);
		
		return baselinePositions.get(ideal) - positionsWithSingleIdeal.get(ideal); 
	}
	
	private Pair<String, Double> cheaterScore(String cheater) {
		return extractRunWithName(cheater, originalScores);
	}
	
	private static Pair<String, Double> extractRunWithName(String runName, List<Pair<String, Double>> runs) {
		List<Pair<String, Double>> ret = runs.stream()
				.filter(i -> runName.equals(i.getKey()))
				.collect(Collectors.toList());
		
		if(ret.size() != 1) {
			throw new RuntimeException("");
		}
		
		return ret.get(0);
	}
	
	private static Map<String, Integer> runToPosition(List<Pair<String, Double>> runs) {
		failIfScoresContainsDuplicatedRuns(runs);
		List<String> ranking = sortedStream(runs)
				.map(Pair::getLeft)
				.collect(Collectors.toList());
		
		return IntStream.range(0, ranking.size())
				.mapToObj(i -> Pair.of(ranking.get(i), i))
				.collect(Collectors.toMap(Pair::getKey, Pair::getValue));
	}
}
