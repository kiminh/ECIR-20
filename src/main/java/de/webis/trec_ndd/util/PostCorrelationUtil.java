package de.webis.trec_ndd.util;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.BiFunction;

import org.apache.commons.math3.stat.correlation.KendallsCorrelation;
import org.apache.commons.math3.stat.correlation.SpearmansCorrelation;
import org.apache.commons.math3.util.Precision;

import java.util.stream.Collectors;

import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;
import lombok.experimental.UtilityClass;

@UtilityClass
public class PostCorrelationUtil
{
	@Data
	@NoArgsConstructor
	@Accessors(chain=true)
	public static class Post {
		String id;
	}
	
	public static void main(String[] args) {
		List<Post> a = Arrays.asList(
			new Post().setId("f"),
			new Post().setId("s"),
			new Post().setId("v"),
			new Post().setId("r"),
			new Post().setId("b")
		);
		
		List<Post> b = Arrays.asList(
			new Post().setId("f"),
			new Post().setId("s"),
			new Post().setId("v"),
			new Post().setId("r"),
			new Post().setId("a")
		);
		
		System.out.println(unionSpearmanCorrelation(a,b));
	}
	
	public static double unionKendallTauCorrelation(List<Post> firstRanking, List<Post> secondRanking)
	{
		List<Post> expandedFirstRanking = expandFirstListWithEntriesFromSecond(firstRanking, secondRanking);
		List<Post> expandedSecondRanking = expandFirstListWithEntriesFromSecond(secondRanking, firstRanking);
		
		return intersectionKendallTauCorrelation(expandedFirstRanking, expandedSecondRanking);
	}
	
	public static double intersectionKendallTauCorrelation(List<Post> firstRanking, List<Post> secondRanking)
	{
		return calculateCorrelation(firstRanking, secondRanking, new KendallsCorrelation()::correlation);
	}
	
	public static double unionSpearmanCorrelation(List<Post> firstRanking, List<Post> secondRanking)
	{
		List<Post> expandedFirstRanking = expandFirstListWithEntriesFromSecond(firstRanking, secondRanking);
		List<Post> expandedSecondRanking = expandFirstListWithEntriesFromSecond(secondRanking, firstRanking);
		
		return intersectionSpearmanCorrelation(expandedFirstRanking, expandedSecondRanking);
	}
	
	public static double intersectionSpearmanCorrelation(List<Post> firstRanking, List<Post> secondRanking)
	{
		return calculateCorrelation(firstRanking, secondRanking, new SpearmansCorrelation()::correlation);
	}
	
	private static double calculateCorrelation(List<Post> firstRanking, List<Post> secondRanking, BiFunction<double[], double[], Double> correlation)
	{
		double[] firstRankingArray = buildCorrelationArrayForFirstRanking(firstRanking, secondRanking);
		double[] secondRankingArray = buildCorrelationArrayForFirstRanking(secondRanking, firstRanking);
		
		if(firstRankingArray.length == 0)
		{
			return 0;
		}
		else if(firstRankingArray.length == 1)
		{
			return Precision.equals(firstRankingArray[0], secondRankingArray[0]) ? 1 : 0;
		}
		
		return correlation.apply(firstRankingArray, secondRankingArray);
	}
	
	private static double[] buildCorrelationArrayForFirstRanking(List<Post> firstRanking, List<Post> secondRanking)
	{
		List<String> firstRankingIds = mapToIds(firstRanking);
		firstRankingIds.retainAll(mapToIds(secondRanking));
		
		List<String> indexValues = new ArrayList<>(firstRankingIds);
		indexValues.sort(String::compareTo);
		
		return indexValues.stream()
				.map(id -> firstRankingIds.indexOf(id))
				.mapToDouble(Integer::doubleValue)
				.toArray();
	}
	
	private static List<Post> expandFirstListWithEntriesFromSecond(List<Post> firstRanking, List<Post> secondRanking)
	{
		List<String> firstRankingIds = mapToIds(firstRanking);
		List<String> secondRankingIds = mapToIds(secondRanking);
		
		secondRankingIds.removeAll(firstRankingIds);
		
		if(secondRankingIds == null || secondRanking == null || secondRankingIds.size() == secondRanking.size())
		{
			return new ArrayList<>();
		}
		
		firstRankingIds.addAll(secondRankingIds);
		
		return firstRankingIds.stream()
				.map(id -> new Post().setId(id))
				.collect(Collectors.toList());
	}
	
	private static List<String> mapToIds(List<Post> posts)
	{
		if(posts == null)
		{
			return new ArrayList<>();
		}
		
		List<String> ret = posts.stream().map(Post::getId).collect(Collectors.toList());
		
		if(ret.stream().anyMatch(a -> a == null || a.isEmpty()) || ret.stream().distinct().collect(Collectors.toList()).size() != ret.size())
		{
			throw new IllegalArgumentException("The following list of posts contains duplicates or missing ids. "
					+ "I dont know how to calculate a correlation coefficient: "+ ret);
		}
		
		return ret;
	}
}
