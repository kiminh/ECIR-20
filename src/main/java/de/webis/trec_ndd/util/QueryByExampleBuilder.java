package de.webis.trec_ndd.util;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;

import org.apache.lucene.analysis.Analyzer;
import org.apache.lucene.analysis.TokenStream;
import org.apache.lucene.analysis.standard.StandardAnalyzer;
import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;

import lombok.SneakyThrows;
import lombok.experimental.UtilityClass;

@UtilityClass
public class QueryByExampleBuilder {
	public static String esQueryByExample(String document){
		return esOrQuery(tokensInText(document));
	}
	
	public static String esQueryByExampleWithKMedianTokens(String document, int k){
		List<String> sortedTokens = new ArrayList<String>(tokensInText(document));
		Collections.sort(sortedTokens, (a,b) -> {
			try {
				return WordCounts.getWordCount(a).compareTo(WordCounts.getWordCount(b));
			} catch (ExecutionException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
				System.exit(1);
				return 1;
			}
		});
		Set<String> ret = new HashSet<>();
		
		while(k > ret.size() && sortedTokens.size() > 0) {
			ret.add(sortedTokens.remove(sortedTokens.size()/2));
		}
		
		return esOrQuery(ret);
	}
	
	private static String esOrQuery(Collection<String> tokens) {
		return tokens.stream()
			.map(t -> "(body_lang.en: \"" + t + "\")")
			.collect(Collectors.joining(" OR "));
	}

	private static Set<String> tokensInText(String text) {
		List<String> ret = textToTokenList(text);
		return new HashSet<>(ret);
	}

	private static List<String> tokensInText(Analyzer analyzer, String text) throws IOException {
		List<String> ret = new LinkedList<>();
		TokenStream tokenStream = analyzer.tokenStream("", text);

		CharTermAttribute attr = tokenStream.addAttribute(CharTermAttribute.class);
		tokenStream.reset();

		while (tokenStream.incrementToken()) {
			ret.add(attr.toString());
		}

		return ret;
	}

	@SneakyThrows
	public static List<String> textToTokenList(String text) {
		Analyzer analyzer = new StandardAnalyzer();
		
		return tokensInText(analyzer, text);
	}
}
