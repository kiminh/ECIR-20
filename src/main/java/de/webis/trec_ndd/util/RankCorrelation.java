package de.webis.trec_ndd.util;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import org.apache.commons.lang3.tuple.Pair;
import org.apache.commons.math3.stat.correlation.KendallsCorrelation;
import org.apache.commons.math3.stat.correlation.SpearmansCorrelation;

import com.google.common.collect.Streams;

import de.webis.trec_ndd.spark.SparkDeduplicateAndEvaluate;
import de.webis.trec_ndd.util.PostCorrelationUtil.Post;
import lombok.SneakyThrows;

public class RankCorrelation {

	public static double kendallTau(List<String> a, List<String> b) {
		Map<String, Double> dictionary = dictionary(a, b);
		List<Post> ap = a.stream().map(i -> new Post().setId(i)).collect(Collectors.toList()); 
		List<Post> bp = b.stream().map(i -> new Post().setId(i)).collect(Collectors.toList());
		
		return PostCorrelationUtil.intersectionKendallTauCorrelation(ap, bp);
//		return new SpearmansCorrelation().correlation(
//			data(a, dictionary),
//			data(b, dictionary)
//		);
	}
	
	private static Map<String, Double> dictionary(List<String> a, List<String> b) {
		Set<String> ret = new HashSet<>(a);
		ret.addAll(b);
		
		return Streams.mapWithIndex(ret.stream().sorted(), (str, i) -> Pair.of(str, Double.valueOf((double)i)))
			.collect(Collectors.toMap(Pair::getKey, Pair::getValue));
	}

	
	private static double[] data(List<String> list, Map<String, Double> dictionary) {
		return list.stream()
				.mapToDouble(dictionary::get)
				.toArray();
	}

	public static double sortListsAndCalculateKendallTau(List<Pair<String, Double>> a, List<Pair<String, Double>> b) {
		return kendallTau(sort(a), sort(b));
	}
	
	private static List<String> sort(List<Pair<String, Double>> ret) {
		return ret.stream()
				.sorted(ScoringChanges.trecStyleComperator())
				.map(Pair::getLeft)
				.collect(Collectors.toList());
	}

	public static double sortedKendallTauAtK(List<Pair<String, Double>> a, List<Pair<String, Double>> b, int k) {
		return kendallTauAtK(sort(a), sort(b), k);
	}
	
	
	public static double kendallTauAtK(List<String> a, List<String> b, int k) {
		List<String> tmpA = a.stream().limit(k).collect(Collectors.toList());
		List<String> tmpB = b.stream().limit(k).collect(Collectors.toList());
		
		tmpA = expandFirstListWithEntriesFromSecond(tmpA, tmpB);
		tmpB = expandFirstListWithEntriesFromSecond(tmpB, tmpA);
		
		return kendallTau(tmpA, tmpB);
	}
	
	@SneakyThrows
	public static void main(String[] args) {
		String bla = "/home/maik/workspace/wstud-thesis-bittner/intermediate-results/gov-evaluation.jsonl";
		List<String>ret = new ArrayList<>();
		
		for(String json : Files.readAllLines(Paths.get(bla))) {
			ret.add(SparkDeduplicateAndEvaluate.fromString(json));
		}
		
		Files.write(Paths.get(bla), ret);
	}
	
	private static List<String> expandFirstListWithEntriesFromSecond(List<String> a, List<String> b)
	{
		List<String> ret = new ArrayList<>(a);
		List<String> tmp = new ArrayList<>(b);
		tmp.removeAll(ret);
		
		ret.addAll(tmp);
		
		return ret;
	}

}
