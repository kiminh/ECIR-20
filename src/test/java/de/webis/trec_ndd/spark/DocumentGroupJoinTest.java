package de.webis.trec_ndd.spark;

import java.util.Arrays;
import java.util.List;

import org.approvaltests.Approvals;
import org.junit.Test;

public class DocumentGroupJoinTest {
	@Test
	public void approveThatTwoDisjountGroupsAreAppended() {
		List<DocumentGroup> contentEquivalent = Arrays.asList(
			new DocumentGroup("{\"hash\":\"a\",\"ids\":[\"a1\",\"a2\",\"a3\"]}"),
			new DocumentGroup("{\"hash\":\"b\",\"ids\":[\"b1\",\"b2\"]}")
		);
		List<DocumentGroup> retrievalEquivalent = Arrays.asList(
			new DocumentGroup("{\"hash\":\"c\",\"ids\":[\"c1\",\"c3\"]}")
		);
		
		Approvals.verifyAsJson(DocumentGroup.appendRetrievalEquivalentToContentEquivalent(contentEquivalent, retrievalEquivalent));
	}
	
	@Test
	public void approveThatRetrievalEquivCanBeAppendedToFirstList() {
		List<DocumentGroup> contentEquivalent = Arrays.asList(
			new DocumentGroup("{\"hash\":\"a\",\"ids\":[\"a1\",\"a2\",\"a3\"]}"),
			new DocumentGroup("{\"hash\":\"b\",\"ids\":[\"b1\",\"b2\"]}")
		);
		List<DocumentGroup> retrievalEquivalent = Arrays.asList(
			new DocumentGroup("{\"hash\":\"c\",\"ids\":[\"c1\",\"c3\",\"a3\"]}")
		);
		
		Approvals.verifyAsJson(DocumentGroup.appendRetrievalEquivalentToContentEquivalent(contentEquivalent, retrievalEquivalent));
	}
	
	@Test
	public void approveThatRetrievalEquivCanBeAppendedToFirstListMultipleTimes() {
		List<DocumentGroup> contentEquivalent = Arrays.asList(
			new DocumentGroup("{\"hash\":\"a\",\"ids\":[\"a1\",\"a2\",\"a3\"]}"),
			new DocumentGroup("{\"hash\":\"b\",\"ids\":[\"b1\",\"b2\"]}")
		);
		List<DocumentGroup> retrievalEquivalent = Arrays.asList(
			new DocumentGroup("{\"hash\":\"c\",\"ids\":[\"c1\",\"c3\",\"a3\"]}"),
			new DocumentGroup("{\"hash\":\"c\",\"ids\":[\"c1\",\"c3\"]}"),
			new DocumentGroup("{\"hash\":\"c\",\"ids\":[\"d1\",\"a3\",\"b1\"]}")
		);
		
		Approvals.verifyAsJson(DocumentGroup.appendRetrievalEquivalentToContentEquivalent(contentEquivalent, retrievalEquivalent));
	}
}
