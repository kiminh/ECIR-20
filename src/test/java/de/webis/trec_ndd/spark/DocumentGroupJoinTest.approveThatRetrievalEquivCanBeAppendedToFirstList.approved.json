[
  {
    "hash": "a",
    "ids": [
      "a1",
      "a2",
      "a3",
      "c1",
      "c3"
    ]
  },
  {
    "hash": "b",
    "ids": [
      "b1",
      "b2"
    ]
  }
]